hs_core/views/__init__.py

```diff
@@ -670,12 +672,22 @@ def _unshare_resource_with_users(request, requesting_user, users_to_unshare_with
     return go_to_resource_listing_page
 
 
-def _set_resource_sharing_status(request, user, resource, is_public):
+def _set_resource_sharing_status(request, user, resource, is_public, is_discoverable=False):
+    if not user.uaccess.can_change_resource_flags(resource):
+        messages.error(request, "You don't have permission to change resource sharing status")
+        return
+
     if is_public and not resource.can_be_public:
         messages.error(request, "Resource may not have sufficient required metadata to be public")
     else:
-        if user.uaccess.can_change_resource_flags(resource):
-            resource.raccess.public = is_public
-            resource.raccess.save()
+        if is_discoverable:
+            if resource.can_be_public:
+                resource.raccess.public = False
```

Comment:
When a resource is set to be discoverable it also needs to be set as private (public=False). The parameter ```is_public``` is not relevant in this case.